### PR TITLE
fix: persist git-diff base spec and re-resolve at diff time (#800)

### LINKED ISSUE #800: Git diff worker: frozen baseCommit diverges from GitHub merge-base diff; persist base spec and re-resolve at diff time
## Problem

The git-diff worker's `baseCommit` is resolved to a commit hash **once at session creation** (`calculateBaseCommit()` → `git merge-base <defaultBranch> HEAD`) and then **frozen and persisted** (`workers.base_commit`). The diff is computed as `git diff <frozenHash>` thereafter.

This means the diff view diverges from GitHub's PR diff (three-dot / merge-base view) as soon as the branch incorporates upstream commits (merge of main, rebase): upstream changes appear as "unexpected file diffs".

## Evidence (reproduced on real data)

Orchestrator session `c0822cc3` (worktree `chore/orchestrator-sprint-2026-04-29`, created 2026-04-28):

| | |
|---|---|
| Frozen `base_commit` in DB | `50238b4` (resolved at session creation) |
| Actual `git merge-base main HEAD` today | `f46fbac` |
| `git diff --stat 50238b4` | **92 files, +5069 / -442** |
| `git diff --stat f46fbac` | **empty (0 files)** |

The branch has zero own changes, yet the Diff tab shows 92 files — all of them the 33 upstream commits absorbed since creation. This is exactly the owner-reported symptom.

## Design (owner-approved)

1. **Stop freezing: persist the base *spec*, not the resolved hash.**
   - Persist the user's intent (e.g., `merge-base:origin/main`, `merge-base:main`, a branch name, or an explicit commit hash) and resolve it **at diff time**, every time.
   - An explicitly chosen commit hash remains pinned (that is the user's stated intent); branch names and `merge-base:` specs re-resolve on each diff.
   - Reuse the existing `MERGE_BASE_REF_PREFIX` (`merge-base:`) spec format from `packages/shared/src/types/git-diff.ts`.

2. **No hidden magic: expose both merge-base variants in `BaseCommitSelector`.**
   - `Fork point from origin/<default>` (merge-base) — shown when `origin/<default>` exists
   - `Fork point from <default>` (local, merge-base)
   - The selector already lists `origin/main` and `main` separately in its "Default branch" section; this extends the same explicitness to the merge-base entries.

3. **Default spec = `merge-base:origin/<default>`**, falling back to `merge-base:<default>` when no remote default exists. Rationale: the feature's promise is "same view as the GitHub PR diff", and in the worktree workflow the local default branch is usually just a stale leftover ref. The default is visible in the UI (not implicit) and switchable to the local variant.

## Boundary behaviors (acceptance criteria)

- `origin/<default>` exists → new git-diff workers default to `merge-base:origin/<default>`
- No remote default branch → fall back to `merge-base:<default>` (local)
- No default branch at all → existing first-commit fallback (`rev-list --max-parents=0 HEAD`) is preserved
- `git merge-base` fails (unrelated histories, deleted ref) → error surfaced to the client (existing `sendError` path), not a silent empty diff
- **Regression test for the symptom:** after merging the default branch into the feature branch, the diff must NOT include upstream-only changes (base re-resolves to the new merge-base)
- After a rebase onto a newer default branch, the diff reflects the new fork point
- Explicit commit-hash base remains pinned across diffs (no re-resolution)

## Migration

Existing persisted `base_commit` values are resolved hashes (mostly stale fork points — i.e., the bug itself). Migrate existing git-diff workers to the new default spec. Note in the migration that explicit user-pinned bases cannot be distinguished from frozen auto-bases and are reset to the default spec.

## Affected code (investigated)

- `packages/server/src/services/git-diff-service.ts` — `calculateBaseCommit()`
- `packages/server/src/services/worker-manager.ts:263-285, 611, 644, 666` — init/persist/restore
- `packages/server/src/websocket/git-diff-handler.ts:156-179` — `set-base-commit` resolution
- `packages/server/src/lib/git.ts` — `getDiff`/`getDiffNumstat`/`getMergeBase`/`getDefaultBranch`
- `packages/client/src/components/workers/BaseCommitSelector.tsx` — selector entries
- DB: `workers.base_commit` column semantics change (hash → spec); migration required

### PR DESCRIPTION
**Pull request by ms2sato**
## Summary

The git-diff worker resolved `baseCommit` to a commit hash **once at session creation** and froze it in `workers.base_commit`. The diff was computed as `git diff <frozenHash>` thereafter, so once the feature branch absorbed upstream commits (merge of main, rebase), those upstream changes showed up as "unexpected" file diffs — diverging from GitHub's three-dot (merge-base) PR view.

This PR persists the base **spec** (intent) instead of the resolved hash and **re-resolves it on every diff computation**.

Closes #800

## What changed

**Resolution model (`packages/shared`, `packages/server`)**
- New `DEFAULT_FORK_POINT_SPEC` sentinel + `resolveBaseSpec()` / `computeDefaultBaseSpec()` in `git-diff-service.ts`.
  - `merge-base:<ref>` and branch-name specs re-resolve every diff; explicit commit-hash specs stay **pinned**.
  - Resolution failure (unrelated histories / deleted ref) surfaces an error via the existing `sendError` / `ValidationError` path — **not** a silent empty diff.
  - `getFirstCommit()` takes only the first line of `rev-list --max-parents=0 HEAD` (repos with merged unrelated histories can emit multiple roots).
- New git-diff workers default to `merge-base:origin/<default>`, falling back to `merge-base:<default>` (local), then the first commit.
- WebSocket handler holds the base **spec** (re-resolved on each send); REST `/diff` and `/diff/file` routes resolve the spec first.
- Branch-rename path keeps the spec unchanged (it re-resolves automatically) instead of freezing a new hash — same fix applied to the **inactive-session** rename path in `session-metadata-service.ts` (sibling call site, bundled per design-principles).

**DB migration**
- `migrateToV21` resets existing git-diff workers' frozen `base_commit` to the sentinel spec. Note: user-pinned explicit bases can't be distinguished from auto-frozen merge-bases at the DB layer, so they're reset to the default spec too (acceptable per the Issue).

**UI (`packages/client`)**
- `BaseCommitSelector` now exposes **both** merge-base variants explicitly — `Fork point from origin/<default>` (shown when the remote default exists, listed first) and `Fork point from <default>` (local) — so the comparison target is never ambiguous.

**Docs**
- `docs/glossary.md`: added `GitDiffWorker` and `Base Spec` entries (glossary-maintenance trigger).

## Boundary behaviors covered (acceptance criteria)

- `origin/<default>` exists → default `merge-base:origin/<default>`; no remote default → `merge-base:<default>`; no default branch → first-commit fallback preserved.
- `git merge-base` failure → error surfaced to the client, not a silent empty diff.
- **Regression (the symptom):** real-repo test — after merging the default branch into the feature branch, the diff does **NOT** include upstream-only changes (base re-resolves to the new fork point). The test also asserts the merge-base actually advanced (`resolved2 !== resolved1`), proving re-resolution happened.
- Explicit commit-hash base remains pinned across diffs; branch names re-resolve to tip.
- Multi-root first-commit edge case; migration v21; selector entry rendering/ordering.

## TDD / polarity

Per `workflow.md`, the real-repo regression test was written first and its polarity verified: against the old frozen-hash model (resolve once, reuse) the upstream-only file **is** included; the shipping fix re-resolves so it is excluded.

## Verification

- `bun run typecheck` → exit 0.
- `bun run test` (full suite) → exit 0 (server 2535 pass / client 1396 pass / shared 324 pass / integration 29 pass / scripts 362 pass).
- CodeRabbit CLI (`--base main`) → **No findings** after addressing the 2 minor multi-root findings.

## Notes / observations

- Manual `set-base-commit` changes remain connection-local (not persisted) — this matches prior behavior and is out of scope for this fix; flagged for potential follow-up.
- The design docs (`session-worker-design.md`, `websocket-protocol.md`) do not document the git-diff worker's base-commit behavior at all (they predate the feature), so there was no frozen-hash spec to update; the new contract is documented in `docs/glossary.md` instead.
- Browser QA of the selector's true-path (both merge-base entries visible) was deferred — it needs a running dev server + a git-diff worker against a repo with both `origin/<default>` and a local default; please dogfood post-merge. Behavior is covered by 5 new unit tests.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by ms2sato** [comment]
Addressed the three CodeRabbit Major findings in `724e58c` (all verified valid):

1. **Sentinel namespace (`git-diff.ts`)** — `DEFAULT_FORK_POINT_SPEC` is now `reserved:default-fork-point`. Git forbids `:` in ref names, so the token can never collide with a real branch/tag, and a real ref literally named `default-fork-point` now round-trips through the persisted base spec (it falls through to `rev-parse` instead of being intercepted). `resolveBaseSpec` special-cases only the exact reserved token. Added a service test asserting the bare `default-fork-point` string resolves via `resolveRef`, not the sentinel path.

2. **Validate-before-replace (`git-diff-handler.ts`)** — `sendDiffData` now returns whether the diff was sent; `set-base-commit` and `updateBaseCommit` commit `state.baseSpec` only after a successful resolve+send. A failed resolution (deleted/invalid ref) retains the last good spec instead of poisoning every later `refresh`/watcher send. Added handler tests: a failed `set-base-commit` keeps the old spec (subsequent `refresh` diffs against the old spec), and `updateBaseCommit` with an unresolvable spec leaves the old spec intact.

3. **Cast-through-`unknown` (`BaseCommitSelector.test.tsx`)** — fetch stub typed as `typeof fetch` and assigned directly; `createBranchesResponse` returns a real `new Response(JSON.stringify(body), …)`. No `as unknown as` remains.

Also fixed the **preflight coverage gate**: added sibling-test cases that exercise the diff in `worker-manager.ts` (`initializeGitDiffWorker` stores a spec, not a resolved hash) and `routes/workers.ts` (`/diff` + `/diff/file` resolve the spec and surface an error on failure). Local `preflight-check.js` now exits 0 with all production files covered.

**Integration test (deferred, with rationale):** preflight flags the cross-package change as "integration test strongly recommended" (advisory ⚠, non-blocking). The `packages/integration` suite has no existing git-diff / WS-worker boundary harness, so a WS+git-repo integration test would require building substantial new infrastructure. The symptom's true path is already covered end-to-end by `git-diff-base-reresolve.test.ts`, which drives the genuine production `resolveBaseSpec` → `getDiffData` pipeline against a **real git repository** through an actual `git merge`; the WS message path (incl. validate-before-replace failure modes) is covered by handler unit tests, and the REST path by route tests. The cross-package contract here is a constant/type contract, not the form/serialization boundary integration tests specifically target. Deferring per the documented allowance.

Verification: `bun run typecheck` exit 0; `bun run test` full suite green; `preflight-check.js` exit 0.

File: packages/shared/src/types/git-diff.ts [hunk 1]
```diff
@@ -169,6 +171,17 @@ export interface ReviewQueueGroup {
...
+/**
+ * Sentinel base spec persisted for migrated git-diff workers and used as a
+ * fallback. Means "resolve the repository's default fork point fresh on each
+ * diff": prefer the merge-base with `origin/<default>`, fall back to the
+ * merge-base with the local `<default>` branch, then to the repository's first
+ * commit. Because it re-resolves on every diff, the base tracks the moving
+ * fork point (e.g. after the feature branch absorbs upstream commits), matching
+ * GitHub's three-dot PR view.
+ */
+export const DEFAULT_FORK_POINT_SPEC = 'default-fork-point';
```
[the remark this answers is not in the document]
Comment (ms2sato) [reply packages/shared/src/types/git-diff.ts]: Resolved in 724e58c: DEFAULT_FORK_POINT_SPEC is now `reserved:default-fork-point` (git forbids `:` in ref names), so real refs round-trip and resolveBaseSpec special-cases only the reserved token. Added a service test asserting a bare `default-fork-point` ref falls through to rev-parse.

File: packages/server/src/websocket/git-diff-handler.ts [hunk 2]
```diff
@@ -150,32 +160,16 @@ export function createGitDiffHandlers(deps: GitDiffHandlerDependencies) {
...
-            } else {
-              sendError(ws, `Invalid ref: ${parsed.ref}`);
-            }
-          }
+          // Store the raw ref as the connection-local base spec. sendDiffData
+          // re-resolves it (merge-base:, branch name, or hash) and surfaces an
+          // error if resolution fails. This change is connection-local and is
+          // not persisted (matching the prior behavior).
+          state.baseSpec = parsed.ref;
+          await sendDiffData(ws, locationPath, state.baseSpec, currentTargetRef);
```
[the remark this answers is not in the document]
Comment (ms2sato) [reply packages/server/src/websocket/git-diff-handler.ts]: Resolved in 724e58c: sendDiffData now returns whether the diff was sent; set-base-commit and updateBaseCommit commit state.baseSpec only after a successful resolve+send (validate-before-replace), so a failed resolution retains the last good spec. Added handler tests covering both failure paths.

File: packages/client/src/components/workers/__tests__/BaseCommitSelector.test.tsx [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+import userEvent from '@testing-library/user-event';
+import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
+import { BaseCommitSelector } from '../BaseCommitSelector';
+import type { BranchesResponse } from '../../../lib/api';
+
+// Mock fetch at the lowest level; fetchSessionBranches goes through the hono
+// client which ultimately calls globalThis.fetch.
+const originalFetch = globalThis.fetch;
+const mockFetch = mock(() => Promise.resolve(new Response()));
+globalThis.fetch = mockFetch as unknown as typeof fetch;
```
[the remark this answers is not in the document]
Comment (ms2sato) [reply packages/client/src/components/workers/__tests__/BaseCommitSelector.test.tsx]: Resolved in 724e58c: removed all cast-through-unknown — fetch stub typed as `typeof fetch` and assigned directly, createBranchesResponse returns a real `new Response(JSON.stringify(body), …)`. grep confirms no `as unknown as` remains.
